apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs

```diff
@@ -246,26 +248,67 @@ public class ReadTests : IClassFixture<TestDatabaseFixture> {
         [Fact]
         public async Task TestSelectOne()
         {
-            Exception exceptionCaught = null;
-            long result = 0;
-            try
-            {
-                using (var connection = await _testFixture.GetTestDatabaseConnectionAsync())
-                {
-                    var cmd =
-                        connection.CreateSelectCommand("SELECT 1");
-                    result = await cmd.ExecuteScalarAsync<long>();
-                }
-            }
-            catch (Exception e)
-            {
-                exceptionCaught = e;
-            }
-
-            Assert.Null(exceptionCaught);
+            long result = await ExecuteAsync<long>("SELECT 1");
             Assert.Equal(1, result);
         }
 
+        [Fact]
+        public async Task TestPositiveInf()
+        {
+            double result = await ExecuteAsync<double>("SELECT IEEE_DIVIDE(1, 0)");
+            Assert.True(double.IsPositiveInfinity(result));
+        }
+
+        [Fact]
+        public async Task TestNegativeInf()
+        {
+            double result = await ExecuteAsync<double>("SELECT IEEE_DIVIDE(-1, 0)");
+            Assert.True(double.IsNegativeInfinity(result));
+        }
+
+        [Fact]
+        public async Task TestNaN()
+        {
+            double result = await ExecuteAsync<double>("SELECT IEEE_DIVIDE(0, 0)");
+            Assert.True(double.IsNaN(result));
+        }
+
+        [Fact]
+        public async Task TestNaNArray()
+        {
+            double[] result = await ExecuteAsync<double[]>("SELECT [IEEE_DIVIDE(1, 0), IEEE_DIVIDE(-1, 0), IEEE_DIVIDE(0, 0)]");
+            Assert.True(result.Length == 3);
+            Assert.True(double.IsPositiveInfinity(result[0]));
+            Assert.True(double.IsNegativeInfinity(result[1]));
+            Assert.True(double.IsNaN(result[2]));
+        }
+
+        [Fact]
+        public async Task TestStructArray()
+        {
+            string sqlQuery = "SELECT ARRAY(SELECT AS STRUCT C1, C2 "
+                              + "FROM (SELECT 'a' AS C1, 1 AS C2 UNION ALL SELECT 'b' AS C1, 2 AS C2) "
+                              + "ORDER BY C1 ASC)";
+            IList result = await ExecuteAsync<IList>(sqlQuery);
+            Assert.True(result.Count == 2);
+            IDictionary s1 = result[0] as IDictionary;
+
+            Assert.True((string)s1["C1"] == "a");
+            Assert.True((long)s1["C2"] == 1);
+
+            s1 = result[1] as IDictionary;
+            Assert.True((string)s1["C1"] == "b");
+            Assert.True((long)s1["C2"] == 2);
+        }
+
+        [Fact]
+        public async Task TestEmptyStructArray()
+        {
+            string sqlQuery = "SELECT ARRAY(SELECT AS STRUCT * FROM (SELECT 'a', 1) WHERE 0 = 1)";
+            IList result = await ExecuteAsync<IList>(sqlQuery);
+            Assert.True(result == null || result.Count == 0);
```

Comment:
ExecuteScalar works by running executereader and looking at the first row.  If there is no first row, it returns default(T) for all T.   An empty collection would change this design.  Let me think how hard it would be to change and lets get user feedback on the alpha.